### PR TITLE
Migration doc for GlobalSettings

### PR DESCRIPTION
**Pull request by dotta**
Migration doc for helping users moving away from `GlobalSettings` in the upcoming Play 2.4 release.

Note that some of the links I used are broken (missing page). For instance, we don't have a `JavaHttpRequestHandlers` page in the doc. I've worked under the assumption that we will create the missing pages before 2.4 final is published. Let me know if I'm wrong.

### CODE REVIEW

File: documentation/manual/releases/Migration24.md [hunk 1]
```diff
@@ -152,6 +153,61 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+* `GlobalSettings.onError`: Create a class that extends [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and move the implementation of your `GlobalSettings.onError` inside the `HttpErrorHandler.onServerError` method. Read [[Error Handling|JavaErrorHandling]] for more information.
+
+* `GlobalSettings.onRequest`: Create a class that extends [`DefaultHttpRequestHandler`](api/java/play/http/DefaultHttpRequestHandler.html), and move the implementation of your `GlobalSettings.onRequest` method inside the `DefaultHttpRequestHandler.createAction` method. Read [[Request Handlers|JavaHttpRequestHandlers]] for more information.
+
+* `GlobalSettings.onRouteRequest`: FIXME: I don't see how this method can be refactored. `JavaCompatibleHttpRequestHandler.routeRequest` is involved, but the injected `JavaCompatibleHttpRequestHandler` instance cannot be replaced. I need help on this one.
```
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: I need help on this one /cc @jroper @richdougherty

File: documentation/manual/releases/Migration24.md [hunk 2]
```diff
@@ -152,6 +153,61 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+* `GlobalSettings.beforeStart` and `GlobalSettings.onStart`:  Anything that needs to happen on start up should now be happening in the constructor of a dependency injected class. A class will perform its initialisation when the dependency injection framework loads it. If you need eager initialisation (for example, because you need to execute some code *before* the application is actually started), use the `eagerly` method on [`Binding`](api/scala/play/api/inject/Binding.html).
+
+* `GlobalSettings.onStop`: Add a dependency to [`ApplicationLifecycle`](api/java/play/inject/ApplicationLifecycle.html) on the class that needs to register a stop hook. Then, move the implementation of your `GlobalSettings.onStop` method inside the `Promise` passed to the `ApplicationLifecycle.addStopHook`. Read [[Stopping/cleaning-up|JavaDependencyInjection#Stopping/cleaning-up]] for more information.
+
+* `GlobalSettings.onError`: Create a class that extends [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and move the implementation of your `GlobalSettings.onError` inside the `HttpErrorHandler.onServerError` method. Read [[Error Handling|JavaErrorHandling]] for more information.
```
Comment (benmccann) [inline documentation/manual/releases/Migration24.md]: This should be "implements" instead of "extends"

File: documentation/manual/releases/Migration24.md [hunk 3]
```diff
@@ -152,6 +153,61 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+* `GlobalSettings.onRequest`: Create a class that extends [`DefaultHttpRequestHandler`](api/java/play/http/DefaultHttpRequestHandler.html), and move the implementation of your `GlobalSettings.onRequest` method inside the `DefaultHttpRequestHandler.createAction` method. Read [[Request Handlers|JavaHttpRequestHandlers]] for more information.
+
+* `GlobalSettings.onRouteRequest`: FIXME: I don't see how this method can be refactored. `JavaCompatibleHttpRequestHandler.routeRequest` is involved, but the injected `JavaCompatibleHttpRequestHandler` instance cannot be replaced. I need help on this one.
+
+* `GlobalSettings.onHandlerNotFound`: Create a class that extends [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide and implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
```
Comment (benmccann) [inline documentation/manual/releases/Migration24.md]: What status codes should you check for? I.E. how do you differentiate between `onHandlerNotFound` and `onBadRequest`?

File: documentation/manual/releases/Migration24.md (hunk 1 above)
Comment (jroper) [inline documentation/manual/releases/Migration24.md]: Implement a custom router.  This method leaks the Scala API a bit, and is not easy, or even really possible to implement in Java, which is why I've left it out.

File: documentation/manual/releases/Migration24.md (hunk 3 above)
Comment (jroper) [inline documentation/manual/releases/Migration24.md]: `onClientError` will be invoked with the status code as an argument... ie, 400 will be passed for bad request, 404 for handler not found.

File: documentation/manual/releases/Migration24.md (hunk 2 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: It's a good point. I'm guessing you are suggesting to make the change only for the Java doc, when a class need to `implements` an interface. However, your suggestion made me think that I'd like to use a broader, language independent, term (since `extends` is not accurate in Scala either). So, I'm going to chance "extends" with "inherits from". WDYT?

File: documentation/manual/releases/Migration24.md (hunk 3 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: Yep, that is why I said "provide an implementation" (and just noticed there is a typo! :-)). I'll see how I can reword it to hint to use the status code.

File: documentation/manual/releases/Migration24.md (hunk 2 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: Using "inherits from" is also more accurate because you may end up extending `DefaultHttpRequestHandler` instead of `HttpErrorHandler`.

File: documentation/manual/releases/Migration24.md (hunk 1 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: I see what you mean. I feel slightly bad because Java here feels 2nd class. I understand the technical reasons, and I know we are working to remove these limitations in the future. I guess the best thing to do here is simply hinting that there is no easy migration for this method.

**Comment by dotta** [comment]
@play-pull-request-validator

**Comment by cchantep** [comment]
Maybe the #4308 can be covered by additional changes in this PR.

**Comment by jroper** [comment]
@dotta What was your plan for this?  We can't merge with broken links, since that will break the build.

Also, a lot of the links to Scala documentation are broken, they should be in the format `/api/scala/index.html#play.api.Foo`, not `/api/scala/play/api/Foo.html`.

**Comment by dotta** [comment]
@jroper If merging this PR will break the build, then we definitely shouldn't merge it. I'll work on the missing doc that this PR requires so that we will can merge this soon. Also, thanks for the feedback on the Scala doc, I'll fix it.

Is it ok if this PR stays open (so that we don't forget about it). Or should it be closed until it's ready to merge?

**Comment by jroper** [comment]
Keep open until ready for merge.

**Comment by richdougherty** [comment]
If you want to do the docs in multiple PRs that's fine too. We can merge the bits that are ready and then have more PRs when you write more docs. Completely up to you how you want to do it though. :)

**Comment by jroper** [comment]
As long as the build passes - our tests test that there are no dead internal links :)

**Comment by richdougherty** [comment]
@play-pull-request-validator, please build again now that link validation for scaladoc has changed.

**Comment by richdougherty** [comment]
That's a weird error in NingConfigSpec:

```
[info] NingConfigSpec
[info] 
[info] NingConfigSpec should
[error]   ! case class defaults must match reference.conf defaults
[error]    null (PropertiesParser.java:60)
[error] com.typesafe.config.impl.PropertiesParser.fromProperties(PropertiesParser.java:60)
[error] com.typesafe.config.impl.Parseable$ParseableProperties.rawParseValue(Parseable.java:860)
[error] com.typesafe.config.impl.Parseable$ParseableProperties.rawParseValue(Parseable.java:842)
[error] com.typesafe.config.impl.Parseable.parseValue(Parseable.java:180)
[error] com.typesafe.config.impl.Parseable.parseValue(Parseable.java:174)
[error] com.typesafe.config.impl.Parseable.parse(Parseable.java:299)
[error] com.typesafe.config.ConfigFactory.parseProperties(ConfigFactory.java:638)
[error] com.typesafe.config.ConfigFactory.parseProperties(ConfigFactory.java:649)
[error] play.api.Configuration$.load(Configuration.scala:44)
[error] play.api.Configuration$.load(Configuration.scala:123)
[error] play.api.Configuration$.load(Configuration.scala:129)
[error] play.api.inject.guice.GuiceApplicationBuilder$$anonfun$$lessinit$greater$default$6$1.apply(GuiceApplicationBuilder.scala:19)
[error] play.api.inject.guice.GuiceApplicationBuilder$$anonfun$$lessinit$greater$default$6$1.apply(GuiceApplicationBuilder.scala:19)
[error] play.api.inject.guice.GuiceApplicationBuilder.injector(GuiceApplicationBuilder.scala:66)
[error] play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:93)
[error] play.api.test.FakeApplication.<init>(Fakes.scala:216)
[error] play.api.test.WithApplication$.$lessinit$greater$default$1(Specs.scala:36)
[error] play.api.libs.ws.ning.NingConfigSpec$$anonfun$6$$anonfun$apply$34$$anon$1.<init>(NingConfigSpec.scala:44)
[error] play.api.libs.ws.ning.NingConfigSpec$$anonfun$6$$anonfun$apply$34.apply(NingConfigSpec.scala:44)
[error] play.api.libs.ws.ning.NingConfigSpec$$anonfun$6$$anonfun$apply$34.apply(NingConfigSpec.scala:44)
```

@play-pull-request-validator, please try again.

**Comment by jroper** [comment]
@play-pull-request-validator

**Comment by richdougherty** [comment]
Closing because of #4458.

**Comment by jroper** [comment]
@dotta This still has invalid links wiki links, and so can't be merged:

https://playframework2.ci.cloudbees.com/job/play-master-PRs/4629/console

**Comment by dotta** [comment]
I've rebase against 2.4.x, but maybe I should have rebased against master(?). Not sure what's the standard way of doing it in Play. Let me know.

**Comment by richdougherty** [comment]
Usually you merge into _master_ then `cherry-pick` onto the `2.4.x` branch. If you don't want to `cherry-pick` directly onto a branch (or don't have access) you can create a second PR against the branch once the first PR has been merged.

**Comment by dotta** [comment]
@richdougherty Thanks Rich. By the way, this PR is (finally!) up for review :+1:

File: documentation/manual/releases/Migration24.md [hunk 4]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+Next follows a method-by-method guide for refactoring your code. Because the APIs are slightly different for Java and Scala, make sure to jump to the appropriate subsection.
+
+> Note: If you haven't yet read about dependency injection in Play, make a point to do it now. Follow the appropriate link to learn about dependency injection in Play with [[Java|JavaDependencyInjection]] or [[Scala|ScalaDependencyInjection]].
+
+#### Java
+
+* `GlobalSettings.beforeStart` and `GlobalSettings.onStart`:  Anything that needs to happen on start up should now be happening in the constructor of a dependency injected class. A class will perform its initialisation when the dependency injection framework loads it. If you need eager initialisation (for example, because you need to execute some code *before* the application is actually started), use the `eagerly` method on [`Binding`](api/java/play/api/inject/Binding.html).
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: There are now docs on eager bindings: https://www.playframework.com/documentation/2.4.0-RC3/JavaDependencyInjection#Eager-bindings

File: documentation/manual/releases/Migration24.md [hunk 5]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+* `GlobalSettings.onRequest`: Create a class that inherits from [`DefaultHttpRequestHandler`](api/java/play/http/DefaultHttpRequestHandler.html), and move the implementation of your `GlobalSettings.onRequest` method inside the `DefaultHttpRequestHandler.createAction` method. Read [[Request Handlers|JavaHttpRequestHandlers]] for more information.
+
+* `GlobalSettings.onRouteRequest`: There is no simple migration for this method when using the Java API. If you need this, you will have to keep your Global class around for a little longer.
+
+* `GlobalSettings.onHandlerNotFound`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide an implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```scala
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: Should this be Java code?

File: documentation/manual/releases/Migration24.md [hunk 6]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+if(statusCode == play.api.http.Status.NOT_FOUND) {
+  // move your implementation of `GlobalSettings.onHandlerNotFound` here
+}
+```
+
+* `GlobalSettings.onBadRequest`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide an implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```scala
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: This should be Java code too, I think.

File: documentation/manual/releases/Migration24.md [hunk 7]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+```
+
+* `GlobalSettings.onLoadConfig`: Specify all configuration in your config file or create your own ApplicationLoader (see [[GuiceApplicationBuilder.loadConfig|JavaDependencyInjection#advanced-extending-the-guiceapplicationloader]]).
+
+* `GlobalSettings.filters`: Create a class that inherits from [`HttpFilters`](api/java/play/http/HttpFilters.html), and provide an implementation for `HttpFilter.filters`. Read [[Http Filters|JavaHttpFilters]] for more information.
+
+
+#### Scala
+
+* `GlobalSettings.beforeStart` and `GlobalSettings.onStart`:  Anything that needs to happen on start up should now be happening in the constructor of a dependency injected class. A class will perform its initialisation when the dependency injection framework loads it. If you need eager initialisation (because you need to execute some code *before* the application is actually started), use the `eagerly` method on [`Binding`](api/scala/index.html#play.api.inject.Binding.html).
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: Link to docs: https://www.playframework.com/documentation/2.4.0-RC3/ScalaDependencyInjection#Eager-bindings

File: documentation/manual/releases/Migration24.md [hunk 8]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+* `GlobalSettings.onError`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and move the implementation of your `GlobalSettings.onError` inside the `HttpErrorHandler.onServerError` method. Read [[Error Handling|JavaErrorHandling]] for more information.
+
+* `GlobalSettings.onRequest`: Create a class that inherits from [`DefaultHttpRequestHandler`](api/java/play/http/DefaultHttpRequestHandler.html), and move the implementation of your `GlobalSettings.onRequest` method inside the `DefaultHttpRequestHandler.createAction` method. Read [[Request Handlers|JavaHttpRequestHandlers]] for more information.
+
+* `GlobalSettings.onRouteRequest`: There is no simple migration for this method when using the Java API. If you need this, you will have to keep your Global class around for a little longer.
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: Can they extend `HttpRequestHandler` or is that only in Scala?

File: documentation/manual/working/javaGuide/advanced/http/JavaHttpRequestHandlers.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+# HTTP Request Handlers
+
+Play abstraction for handling requests is less sophisticated in the Play Java API, when compared to its Scala counterpart. However, it may still be enough if you only need to execute some code before the controller's action method associated to the passed request is executed. If that's not enough, then you should fall back to the [[Scala API|ScalaHttpRequestHandlers]] for creating a HTTP request handler.
+
+## Implementing a custom request handler
+
+The [`HttpRequestHandler`](api/java/play/http/HttpRequestHandler) interface has two methods that needs to be implemented: 
+
+* `createAction`: Takes the request and the controller's action method associated with the passed request.
+*  `wrapAction`: Takes the action to be run and allows for a final global intereceptor to be added to the action.
```
Comment (richdougherty) [inline documentation/manual/working/javaGuide/advanced/http/JavaHttpRequestHandlers.md]: s/inter**e**ceptor/interceptor/

**Comment by richdougherty** [comment]
Great work @dotta! A few more changes, but nearly there now…

File: documentation/manual/releases/Migration24.md (hunk 8 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: That's only valid in Scala :-( see https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/http/HttpRequestHandler.scala#L42

**Comment by dotta** [comment]
@richdougherty I've addressed all your comments (by the way, thanks!). Let me know if it looks good now, so that I can squash commits (as it makes sense), and we can finally merge this baby :-)

File: documentation/manual/releases/Migration24.md [hunk 10]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+
+* `GlobalSettings.onHandlerNotFound`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide an implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```java
+if(statusCode == play.api.http.Status.NOT_FOUND) {
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: `play.api` is the Scala API, so I think this should be `play.mvc.Http.Status.NOT_FOUND`.

File: documentation/manual/releases/Migration24.md [hunk 11]
```diff
@@ -201,6 +202,89 @@ While Play 2.4 won't force you to use the dependency injected versions of compon
...
+  // move your implementation of `GlobalSettings.onHandlerNotFound` here
+}
+```
+
+* `GlobalSettings.onBadRequest`: Create a class that inherits from [`HttpErrorHandler`](api/java/play/http/HttpErrorHandler.html), and provide an implementation for `HttpErrorHandler.onClientError`. Read [[Error Handling|JavaErrorHandling]] for more information.
+Note that `HttpErrorHandler.onClientError` takes a `statusCode` in argument, hence your implementation should boil down to:
+
+```java
+if(statusCode == play.api.http.Status.BAD_REQUEST) {
```
Comment (richdougherty) [inline documentation/manual/releases/Migration24.md]: Same here: `play.mvc.Http.Status.BAD_REQUEST`.

**Comment by richdougherty** [comment]
@dotta, just two small changes, then if you squash the commits I'd be happy to merge this. :)

File: documentation/manual/releases/Migration24.md (hunk 10 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: I didn't realise we have two places with the constants. I'll make the change, but I wonder why keeping things this way. Would it be bad to deprecate `play.api.http.Status`?

One more thing, it looks like that  `play.api.http.Status` and `play.mvc.Http.Status` don't actually contain exactly the same constants. Is this an oversight due to maintaining the same bits in two different places?

**Comment by dotta** [comment]
@richdougherty All changes done. I've also rebased against master and squashed commits as it made sense. Feel free to merge when the PR validator has finished running. By the way, could you have a look at [this comment](https://github.com/playframework/playframework/pull/4308#discussion_r30570045) and let me know what you think.

File: documentation/manual/releases/Migration24.md (hunk 10 above)
Comment (dotta) [inline documentation/manual/releases/Migration24.md]: All good now. No need to comment.

**Comment by dotta** [comment]
Alright, finally ready to merge, together with the [backport for 2.4.x](https://github.com/playframework/playframework/pull/4500)
